File: backend/src/Modules/Catalog/Catalog.Domain/Brands/Brand.cs
using SharedKernel;

namespace Catalog.Domain.Brands;

/// <summary>
/// Ürünlerin bağlanabileceği düz (hiyerarşisiz) markayı yöneten kök varlık.
/// </summary>
public sealed class Brand : AggregateRoot<Guid>
{
    private Brand()
    {
    }

    private Brand(Guid id, string name, string? code)
        : base(id)
    {
        Name = name;
        Code = code;
    }

    /// <summary>Gets markanın görünen adı.</summary>
    public string Name { get; private set; } = string.Empty;

    /// <summary>Gets markanın opsiyonel kodu (ör. pazaryeri marka kimliği).</summary>
    public string? Code { get; private set; }

    /// <summary>Yeni marka oluşturur.</summary>
    /// <param name="name">Marka adı.</param>
    /// <param name="code">Opsiyonel marka kodu.</param>
    public static Result<Brand> Create(string name, string? code)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return Result.Failure<Brand>(Error.Validation("Brand name is required."));
        }

        var brand = new Brand(
            Guid.NewGuid(),
            name.Trim(),
            string.IsNullOrWhiteSpace(code) ? null : code.Trim());

        return Result.Success(brand);
    }

    /// <summary>Marka adını ve opsiyonel kodunu günceller.</summary>
    /// <param name="name">Yeni marka adı.</param>
    /// <param name="code">Yeni marka kodu; boş bırakılırsa null olur.</param>
    public Result Rename(string name, string? code)
    {
        if (string.IsNullOrWhiteSpace(name))
        {
            return Result.Failure(Error.Validation("Brand name is required."));
        }

        Name = name.Trim();
        Code = string.IsNullOrWhiteSpace(code) ? null : code.Trim();
        return Result.Success();
    }
}

File: backend-go/internal/modules/catalog/domain/brands/brand.go
// Package brands, ürünlerin bağlanabileceği düz (hiyerarşisiz) marka kök
// varlığını içerir (.NET Catalog.Domain.Brands karşılığı).
package brands

import (
	"strings"

	"github.com/google/uuid"

	"pimly.commerslab/backend-go/internal/sharedkernel"
)

// Brand, markayı yöneten kök varlıktır. Code, opsiyonel marka kodudur
// (ör. pazaryeri marka kimliği); boş dizgi yerine nil saklanır.
type Brand struct {
	// ID, markanın benzersiz kimliğidir.
	ID uuid.UUID

	// Name, markanın görünen adıdır (en çok 500 karakter).
	Name string

	// Code, markanın opsiyonel kodudur; yokluğu nil ile temsil edilir.
	Code *string
}

// NewBrand, doğrulanmış yeni bir marka oluşturur; ad ve kod kırpılır,
// boş kod nil'e çevrilir. Hata mesajı .NET karşılığıyla birebir aynıdır.
func NewBrand(name string, code *string) sharedkernel.ResultOf[*Brand] {
	if strings.TrimSpace(name) == "" {
		return sharedkernel.FailOf[*Brand](sharedkernel.NewValidationError("Brand name is required."))
	}
	return sharedkernel.OkOf(&Brand{
		ID:   uuid.New(),
		Name: strings.TrimSpace(name),
		Code: normalizeCode(code),
	})
}

// Rename, marka adını ve opsiyonel kodunu günceller.
func (b *Brand) Rename(name string, code *string) sharedkernel.Result {
	if strings.TrimSpace(name) == "" {
		return sharedkernel.Fail(sharedkernel.NewValidationError("Brand name is required."))
	}
	b.Name = strings.TrimSpace(name)
	b.Code = normalizeCode(code)
	return sharedkernel.Ok()
}

// normalizeCode, kodu kırpar; boş/beyaz boşluk kodu nil'e çevirir.
func normalizeCode(code *string) *string {
	if code == nil {
		return nil
	}
	trimmed := strings.TrimSpace(*code)
	if trimmed == "" {
		return nil
	}
	return &trimmed
}
